---
title: "コントローラーのテストで例外が発生してテストが失敗する"
draft: false
tags: ["rails","ruby","rspec"]
private: false
slug: "e848d251-d646-4d69-b394-c113fb84410c"
date: "2020-01-27T17:16:30+09:00"
lastmod: "2020-01-27T17:17:08+09:00"
keywords: ["rails","ruby","rspec","ベジプロ","プログ","プログラム"]
# headless: true
---

# 解決
`context` に、 `show_exceptions: true` オプションを設定する。

```rb
context "failed", show_exceptions: true do
  ...
end
```

> [Rails アプリケーションの不安定なテストを撲滅したい 〜system spec のデバッグ方法とテストを不安定にさせる要因〜 - あらびき日記](https://abicky.net/2019/09/17/062506/)

# 事象
APIのレスポンスチェックのテストを作成したが、 `api_get` を実行した時点で例外が発生してテストが失敗する。

## テスト
```rb
subject { api_get "/path/to/api" }
context "failed" do
  before do
    # something for failed
  end

  example "XXXX のため、エラー" do
    subject
    expect(response).to have_http_status(:forbidden)
  end
end
```

## エラー内容
```
Failure/Error: raise ForbiddenError

HogeController::ForbiddenError:
 アクセス権限がありません
# ./app/controllers/hoge_controller.rb:88:in `response_forbidden'
# ./app/controllers/hoge_controller.rb:51:in `block_foreign_hosts'
...
```
